test/integration/cds_integration_test.cc

```diff
@@ -0,0 +1,211 @@
+    if (!(expected_version == discovery_request.version_info())) {
+      return AssertionFailure() << fmt::format("version {} does not match expected {} in {}",
+                                               discovery_request.version_info(), expected_version,
+                                               discovery_request.DebugString());
+    }
+    return AssertionSuccess();
+  }
+
+  template <class T>
+  void sendDiscoveryResponse(const std::string& type_url, const std::vector<T>& messages,
+                             const std::string& version) {
+    envoy::api::v2::DiscoveryResponse discovery_response;
+    discovery_response.set_version_info(version);
+    discovery_response.set_type_url(type_url);
+    for (const auto& message : messages) {
+      discovery_response.add_resources()->PackFrom(message);
+    }
+    cds_stream_->sendGrpcMessage(discovery_response);
+  }
+
+  envoy::api::v2::Cluster buildCluster(const std::string& name) {
+    return TestUtility::parseYaml<envoy::api::v2::Cluster>(
+        fmt::format(R"EOF(
+      name: {}
+      connect_timeout: 5s
+      type: STATIC
+      load_assignment:
+        cluster_name: {}
+        endpoints:
+        - lb_endpoints:
+          - endpoint:
+              address:
+                socket_address:
+                  address: {}
+                  port_value: {}
+      lb_policy: ROUND_ROBIN
+      http2_protocol_options: {{}}
+    )EOF",
+                    name, name, Network::Test::getLoopbackAddressString(ipVersion()),
+                    fake_upstreams_[0]->localAddress()->ip()->port()));
+  }
+
+  void initializeCds() {
+    config_helper_.addConfigModifier([this](envoy::config::bootstrap::v2::Bootstrap& bootstrap) {
+      // Setup cds and corresponding gRPC cluster.
+      auto* cds_config =
+          bootstrap.mutable_dynamic_resources()->mutable_cds_config()->mutable_api_config_source();
+      bootstrap.mutable_dynamic_resources()->clear_ads_config();
+      bootstrap.mutable_dynamic_resources()->clear_lds_config();
+
+      cds_config->set_api_type(envoy::api::v2::core::ApiConfigSource::GRPC);
+      cds_config->mutable_request_timeout()->set_seconds(1);
+      auto* grpc_service = cds_config->add_grpc_services();
+      setGrpcService(*grpc_service, "my_cds_cluster", fake_upstreams_[0]->localAddress());
+      grpc_service->mutable_envoy_grpc()->set_cluster_name("my_cds_cluster");
+
+      auto* cds_cluster = bootstrap.mutable_static_resources()->add_clusters();
+      cds_cluster->set_name("my_cds_cluster");
+      cds_cluster->mutable_connect_timeout()->set_seconds(5);
+      auto* sockaddr = cds_cluster->add_hosts()->mutable_socket_address();
+      sockaddr->set_protocol(envoy::api::v2::core::SocketAddress::TCP);
+      sockaddr->set_address(Network::Test::getLoopbackAddressString(ipVersion()));
+      sockaddr->set_port_value(0);
+      cds_cluster->clear_http2_protocol_options();
+    });
+    setUpstreamProtocol(FakeHttpConnection::Type::HTTP2);
```

Comment:
Can you comment on which upstream this is? In the line above, the H2 options are cleared for the dataplane upstream.

Reply:
It is for the CDS upstream, which is where the test HTTP request is sent. It looks like that clear is unnecessary.